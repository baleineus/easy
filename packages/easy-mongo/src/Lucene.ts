import {
  DateTime,
  entries,
  Func,
  ifDefined,
  ifNotEmpty,
  ifTrue,
  isDefined,
  isEmptyObject,
  isFunction,
  List,
  on,
  OneOrMore,
  Optional,
  RequireAtLeastOne,
  toArray,
  toList,
} from '@thisisagile/easy';
import { toMongoType } from './Utils';

type FuzzyOptions = {
  maxEdits: number;
  prefixLength: number;
  maxExpansions: number;
};

export type Facet =
  | { path: string; type: 'string'; numBuckets: number }
  | { path: string; type: 'number'; boundaries: number[]; default?: string }
  | { path: string; type: 'date'; boundaries: Date[]; default?: string };

export type Operator = Func<Optional<object>, string>;
export type Clause = object | Operator;
export type Clauses = Record<string, Clause>;
export type SearchDefinition = Record<
  string,
  (
    v: string | number,
    q?: Record<string, string | number>
  ) => RequireAtLeastOne<{ should?: Clauses; filter?: Clauses; must?: Clauses; mustNot?: Clauses; sort?: Record<string, 1 | -1>; facet?: Facet }>
>;

type Compound = {
  must: OneOrMore<Clauses>;
  should: OneOrMore<Clauses>;
  mustNot: OneOrMore<Clauses>;
  filter: OneOrMore<Clauses>;
};

const should = (query: Record<string, string | number>, def: SearchDefinition): Clauses[] => entries(query).mapDefined(([k, v]) => def[k]?.(v, query)?.should);
const must = (query: Record<string, string | number>, def: SearchDefinition): Clauses[] => entries(query).mapDefined(([k, v]) => def[k]?.(v, query)?.must);
const mustNot = (query: Record<string, string | number>, def: SearchDefinition): Clauses[] =>
  entries(query).mapDefined(([k, v]) => def[k]?.(v, query)?.mustNot);
const filter = (query: Record<string, string | number>, def: SearchDefinition): Clauses[] => entries(query).mapDefined(([k, v]) => def[k]?.(v, query)?.filter);

export const lucene = {
  clause: (c: Clauses) => entries(c).reduce((res, [k, v]) => res.add(isFunction(v) ? v(k) : v), toList()),
  clauses: (cs: OneOrMore<Clauses>) => toArray(cs).flatMap(c => lucene.clause(c)),
  compound: (query: Record<string, string | number>, def: SearchDefinition): Partial<Compound> =>
    entries({
      should: ifNotEmpty(should(query, def).concat(filter(query, def), must(query, def), mustNot(query, def)), should(query, def), [
        { wildcard: lucene.wildcard() },
      ]),
      ...ifNotEmpty(filter(query, def), f => ({ filter: f })),
      ...ifNotEmpty(mustNot(query, def), m => ({ mustNot: m })),
      ...ifNotEmpty(must(query, def), m => ({ must: m })),
    }).reduce((res, [k, v]) => on(res, r => (r[k] = lucene.clauses(v))), {
      minimumShouldMatch: should(query, def).length > 0 ? 1 : 0,
    } as any),
  search: (c: Partial<Compound>, index?: string) => ({
    $search: {
      ...ifDefined(index, { index }),
      compound: entries(c).reduce((res, [k, v]) => on(res, r => (r[k] = lucene.clauses(v))), {} as any),
    },
  }),
  searchWithDef: (query: Record<string, string | number>, options: SearchDefinition, count: 'total' | 'lowerBound' = 'total', index?: string) => {
    const sort = entries(query)
      .mapDefined(([k, v]) => options[k]?.(v, query)?.sort)
      .first();
    return {
      $search: { ...ifDefined(index, { index }), compound: lucene.compound(query, options), ...ifDefined(sort, { sort }), count: { type: count } },
    };
  },
  searchMeta: (query: Record<string, string | number>, def: SearchDefinition, count: 'total' | 'lowerBound' = 'total', index?: string) => ({
    $searchMeta: {
      ...ifDefined(index, { index }),
      ...ifTrue(
        !isEmptyObject(lucene.facets(def)),
        {
          facet: {
            operator: {
              compound: lucene.compound(query, def),
            },
            facets: lucene.facets(def),
          },
        },
        { compound: lucene.compound(query, def) }
      ),
      count: { type: count },
    },
  }),
  exists: (): Operator => (path: string) => ({ exists: { path } }),
  text:
    (value?: OneOrMore<unknown>, fuzzy?: Partial<FuzzyOptions>): Operator =>
    (path: string) =>
      ifDefined(value, v => ({
        text: {
          path: path === 'wildcard' ? { wildcard: '*' } : path,
          query: v,
          ...ifDefined(fuzzy, { fuzzy }),
        },
      })),
  wildcard:
    (value?: OneOrMore<unknown>, allowAnalyzedField = true): Operator =>
    (path: string) => ({
      wildcard: {
        path: path === 'wildcard' ? { wildcard: '*' } : path,
        query: ifDefined(value, value, '*'),
        allowAnalyzedField,
      },
    }),
  lt:
    (value: unknown): Operator =>
    (path: string) =>
      ifDefined(value, lt => ({ range: { path, lt } })),
  lte:
    (value: unknown): Operator =>
    (path: string) =>
      ifDefined(value, lte => ({ range: { path, lte } })),
  gt:
    (value: unknown): Operator =>
    (path: string) =>
      ifDefined(value, gt => ({ range: { path, gt } })),
  gte:
    (value: unknown): Operator =>
    (path: string) =>
      ifDefined(value, gte => ({ range: { path, gte } })),
  after: (date: unknown): Operator => lucene.gte(toMongoType(date)),
  before: (date: unknown): Operator => lucene.lt(toMongoType(date)),
  between:
    (after: unknown, before: unknown): Operator =>
    (path: string) => ({
      range: {
        path,
        gte: toMongoType(after),
        lt: toMongoType(before),
      },
    }),
  facets: (def: SearchDefinition) =>
    entries(def)
      .filter(([k, v]) => isDefined(v(k)?.facet))
      .map(([k, v]) => ({ [k]: v(k)?.facet }))
      .reduce((acc, v) => ({ ...acc, ...v }), {}),
  facet: {
    string: (path: string, numBuckets = 1000): Facet => ({
      type: 'string',
      path,
      numBuckets,
    }),
    number: (path: string, boundaries: List<number>, alt?: string): Facet => ({
      type: 'number',
      path,
      boundaries,
      ...ifDefined(alt, a => ({ default: a })),
    }),
    date: (path: string, boundaries: List<DateTime>, alt?: string): Facet => ({
      type: 'date',
      path,
      boundaries: boundaries.mapDefined(b => b.toDate()),
      ...ifDefined(alt, a => ({ default: a })),
    }),
  },
};
