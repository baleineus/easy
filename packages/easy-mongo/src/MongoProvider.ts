import {
  asJson,
  asNumber,
  asString,
  choose,
  Condition,
  Database,
  Exception,
  FetchOptions,
  Field,
  Id,
  ifTrue,
  isArray,
  isDefined,
  isField,
  isSortCondition,
  json,
  Json,
  JsonValue,
  LogicalCondition,
  OneOrMore,
  PageList,
  reject,
  Sort,
  toArray,
  toPageList,
  tuple2,
  tuple3,
  when,
} from '@thisisagile/easy';
import {
  AggregationCursor,
  Collection as MongoCollection,
  CreateIndexesOptions,
  Document,
  FindCursor,
  FindOptions as MongoFindOptions,
  IndexSpecification,
  MongoClient,
  StrictFilter as MongoFilter,
} from 'mongodb';
import { Collection } from './Collection';
import { toMongoType } from './Utils';

const omitId = (j: Json): Json => json.delete(j, '_id');

export type Projection = Record<string, 0 | 1>;
export type FindOptions = FetchOptions & { projection?: Projection };
export type Filter<T = unknown> = MongoFilter<T>;
export type Query = Condition | LogicalCondition | Filter<any>;

export type IndexOptions = {
  unique?: boolean;
  filter?: Query;
  languageOverride?: string;
  languageDefault?: string;
};

export type Indexes = OneOrMore<string | Field | Sort | Record<string, 1 | -1>>;

export class MongoProvider {
  protected static readonly clients: { [key: string]: Promise<MongoClient> } = {};

  constructor(readonly coll: Collection, protected client?: Promise<MongoClient>) {}

  static client(db: Database): Promise<MongoClient> {
    return when(db.options?.cluster)
      .not.isDefined.reject(Exception.IsNotValid.because('Missing cluster in database options.'))
      .then(
        u =>
          (MongoProvider.clients[u] =
            MongoProvider.clients[u] ??
            MongoClient.connect(u, {
              auth: {
                username: asString(db.options?.user),
                password: asString(db.options?.password),
              },
              ...(db.options?.maxPoolSize && { maxPoolSize: db.options?.maxPoolSize }),
              ...(db.options?.minPoolSize && { minPoolSize: db.options?.minPoolSize }),
              ...(db.options?.maxIdleTimeMS && { maxIdleTimeMS: db.options?.maxIdleTimeMS }),
            }))
      );
  }

  cluster(): Promise<MongoClient> {
    return Promise.resolve()
      .then(() => this.client ?? (this.client = MongoProvider.client(this.coll.db)))
      .catch(e => {
        this.client = undefined;
        return reject(e);
      });
  }

  toMongoJson(query: Query): Json {
    return toMongoType(asJson(query));
  }

  find(query: Query, options?: FindOptions): Promise<PageList<Json>> {
    return tuple3(this.collection(), this.toMongoJson(query), this.toFindOptions(options))
      .then(([c, q, o]) =>
        tuple2(
          c.find(q, o),
          ifTrue(o.total, () => c.countDocuments(q))
        )
      )
      .then(([res, total]) => this.toArray(res, { ...options, total }));
  }

  all(options?: FindOptions): Promise<PageList<Json>> {
    return this.find({}, options);
  }

  byId(id: Id, options?: FindOptions): Promise<Json> {
    return this.collection().then(c => c.findOne(this.toMongoJson({ id: id }), this.toFindOptions(options)) as Promise<Json>);
  }

  by(key: string, value: JsonValue, options?: FindOptions): Promise<PageList<Json>> {
    return this.find({ [key]: value }, options);
  }

  group(qs: Filter<any>[]): Promise<PageList<Json>> {
    return this.aggregate(qs);
  }

  aggregate(qs: Filter<any>[]): Promise<PageList<Json>> {
    return this.collection()
      .then(c => c.aggregate(qs.map(q => this.toMongoJson(q))))
      .then(res => this.toArray(res));
  }

  add(item: Json): Promise<Json> {
    return this.collection()
      .then(c => c.insertOne(omitId(item)))
      .then(() => omitId(item));
  }

  update(item: Json): Promise<Json> {
    return this.collection()
      .then(c => c.updateOne(this.toMongoJson({ id: item.id }), { $set: omitId(item) }))
      .then(() => this.byId(item.id as Id));
  }

  remove(id: Id): Promise<boolean> {
    return this.collection()
      .then(c => c.deleteOne(this.toMongoJson({ id })))
      .then(d => d.acknowledged);
  }

  count(query?: Query): Promise<number> {
    return this.collection().then(c => c.countDocuments(this.toMongoJson(query ?? {})));
  }

  createIndex(indexes: Indexes, options?: IndexOptions): Promise<string> {
    return this.collection().then(c => c.createIndex(this.toIndexSpecification(indexes), this.toCreateIndexesOptions(options)));
  }

  createPartialIndex(indexes: Indexes, filter: Query, options?: Omit<IndexOptions, 'filter'>): Promise<string> {
    return this.createIndex(indexes, { ...options, filter });
  }

  createTextIndex(indexes: OneOrMore<Field | string>, options?: IndexOptions): Promise<string> {
    const ii = toArray(indexes).reduce((i, f) => ({ ...i, [asString(f)]: 'text' }), {});
    return this.createIndex(ii, { unique: false, ...options });
  }

  collection(): Promise<MongoCollection> {
    return this.cluster()
      .then(c => c.db(this.coll.db.name))
      .then(db => db.collection(asString(this.coll)));
  }

  protected toFindOptions(options?: FindOptions): MongoFindOptions & { total: boolean } {
    return {
      limit: asNumber(options?.take ?? 250),
      ...(options?.skip && { skip: asNumber(options?.skip) }),
      ...((options?.sorts && { sort: options?.sorts }) || (options?.sort && { sort: this.coll.sort(...(options?.sort ?? [])) })),
      total: isDefined(options?.skip) || isDefined(options?.take),
      projection: options?.projection ?? { _id: 0 },
    };
  }

  protected toIndexSpecification(index: Indexes): IndexSpecification {
    return choose(index)
      .type(isField, f => f.property as IndexSpecification)
      .type(isSortCondition, s => s.toJSON() as IndexSpecification)
      .type(isArray, aa => aa.map(a => this.toIndexSpecification(a)) as IndexSpecification)
      .else(i => i as IndexSpecification);
  }

  protected toCreateIndexesOptions(options?: IndexOptions): CreateIndexesOptions {
    return {
      unique: options?.unique ?? true,
      ...(options?.languageOverride && { language_override: options.languageOverride }),
      ...(options?.languageDefault && { default_language: options.languageDefault }),
      ...(options?.filter && { partialFilterExpression: toMongoType(asJson(options.filter)) }),
    };
  }

  protected toArray(
    cursor: FindCursor<Document> | AggregationCursor<Document>,
    options?: { take?: number; skip?: number; total?: number }
  ): Promise<PageList<Json>> {
    return cursor.toArray().then(r => toPageList<Json>(r, options));
  }
}
