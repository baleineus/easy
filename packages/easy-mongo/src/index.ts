export * from './Collection';
export * from './Lucene';
export * from './MongoGateway';
export * from './MongoProvider';
export * from './Stages';
export * from './Utils';
