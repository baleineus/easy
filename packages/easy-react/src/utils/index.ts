export * from './Hooks';
export * from './useOnce';
