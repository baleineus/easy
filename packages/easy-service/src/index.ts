export * from './health/HealthResource';
export * from './health/HealthUri';
export * from './http/OriginatedError';
export * from './http/Verb';
export * from './resources/AppProvider';
export * from './resources/Requires';
export * from './resources/Resource';
export * from './resources/Route';
export * from './resources/Service';
export * from './security/Jwt';
