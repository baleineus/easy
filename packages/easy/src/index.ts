export * from './data/Condition';
export * from './data/InMemoryGateway';
export * from './data/Database';
export * from './data/DataProvider';
export * from './data/Field';
export * from './data/Query';
export * from './data/QueryProvider';

export * from './domain/Audit';
export * from './domain/Child';
export * from './domain/DateTime';
export * from './domain/Entity';
export * from './domain/Struct';
export * from './domain/Repo';
export * from './domain/Typo';

export * from './http/CacheControl';
export * from './http/ContentType';
export * from './http/HttpHeader';
export * from './http/HttpStatus';
export * from './http/HttpVerb';
export * from './http/Request';
export * from './http/RequestOptions';
export * from './http/RequestProvider';
export * from './http/Response';
export * from './http/RestResult';
export * from './http/toResponse';

export * from './process/App';
export * from './process/Manage';
export * from './process/Scope';
export * from './process/Search';
export * from './process/UseCase';

export * from './resources/Req';

export * from './security/Jwt';

export * from './services/Api';
export * from './services/ApiGateway';
export * from './services/AxiosProvider';
export * from './services/MappedRouteGateway';
export * from './services/RouteGateway';
export * from './services/ViewRouteGateway';

export * from './sql/Clause';
export * from './sql/Column';
export * from './sql/Delete';
export * from './sql/Insert';
export * from './sql/Join';
export * from './sql/SqlQuery';
export * from './sql/Select';
export * from './sql/Table';
export * from './sql/TableGateway';
export * from './sql/Update';

export * from './types/AnyKey';
export * from './types/Array';
export * from './types/Builder';
export * from './types/Cache';
export * from './types/CacheAge';
export * from './types/Case';
export * from './types/Constructor';
export * from './types/Context';
export * from './types/Currency';
export * from './types/Enum';
export * from './types/Environment';
export * from './types/ErrorOrigin';
export * from './types/Exception';
export * from './types/Falsy';
export * from './types/Func';
export * from './types/Gateway';
export * from './types/Get';
export * from './types/Id';
export * from './types/Identity';
export * from './types/Is';
export * from './types/IsA';
export * from './types/IsDate';
export * from './types/IsEqual';
export * from './types/Json';
export * from './types/List';
export * from './types/Message';
export * from './types/Meta';
export * from './types/Nullish';
export * from './types/Number';
export * from './types/Object';
export * from './types/PageList';
export * from './types/Parser';
export * from './types/Primitive';
export * from './types/Repository';
export * from './types/Result';
export * from './types/Results';
export * from './types/Sort';
export * from './types/Tag';
export * from './types/Template';
export * from './types/Text';
export * from './types/Try';
export * from './types/TypeGuard';
export * from './types/Types';
export * from './types/Uri';
export * from './types/Uuid';
export * from './types/Validatable';
export * from './types/Value';

export * from './utils/Base64';
export * from './utils/Convert';
export * from './utils/If';
export * from './utils/Loading';
export * from './utils/Log';
export * from './utils/Mapper';
export * from './utils/Promise';
export * from './utils/Property';
export * from './utils/Seconds';
export * from './utils/State';
export * from './utils/Sentence';
export * from './utils/Traverse';
export * from './utils/View';
export * from './utils/Wait';

export * from './validation/Contraints';
export * from './validation/Validate';
export * from './validation/When';
