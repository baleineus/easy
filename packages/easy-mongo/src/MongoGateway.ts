import { Filter, FindOptions, MongoProvider } from './MongoProvider';
import {
  asJson,
  Condition,
  Gateway,
  Id,
  ifDefined,
  isDefined,
  isPresent,
  Json,
  JsonValue,
  List,
  LogicalCondition,
  Optional,
  PageList,
  toArray,
} from '@thisisagile/easy';
import { Collection } from './Collection';
import { stages } from './Stages';

export class MongoGateway implements Gateway<FindOptions> {
  constructor(readonly collection: Collection, readonly provider: MongoProvider = collection.provider) {}

  all(options?: FindOptions): Promise<PageList<Json>> {
    return this.provider.all(options).then(js => js.map(j => this.collection.in(j)));
  }

  byId(id: Id): Promise<Json | undefined> {
    return this.provider.byId(id).then(j => ifDefined(j, this.collection.in(j)));
  }

  by(key: string, value: JsonValue, options?: FindOptions): Promise<PageList<Json>> {
    return this.provider.by(key, value, options).then(js => js.map(j => this.collection.in(j)));
  }

  byIds(...ids: Id[]): Promise<List<Json>> {
    return this.find(this.collection.id.isIn(...ids));
  }

  find(q: JsonValue | Condition | LogicalCondition, options?: FindOptions): Promise<PageList<Json>> {
    return this.provider.find(asJson(q), options).then(js => js.map(j => this.collection.in(j)));
  }

  search(q: JsonValue, options?: FindOptions): Promise<PageList<Json>> {
    return this.find(this.collection.google(q), options);
  }

  filter(options?: FindOptions): Promise<PageList<Json>> {
    return this.all(options);
  }

  exists(id: Id): Promise<boolean> {
    return this.provider.byId(id).then(i => isDefined(i));
  }

  aggregate(...filters: Optional<Filter>[]): Promise<PageList<Json>> {
    return this.provider.aggregate(toArray(...filters).filter(isPresent) as Filter[]);
  }

  match(f: Filter): Promise<PageList<Json>> {
    return this.aggregate(stages.match.match(f));
  }

  add(item: Json): Promise<Json> {
    return this.provider.add(this.collection.out(item)).then(j => this.collection.in(j));
  }

  update(item: Json): Promise<Json> {
    return this.provider.update(this.collection.out(item)).then(j => this.collection.in(j));
  }

  remove(id: Id): Promise<boolean> {
    return this.provider.remove(id);
  }
}
