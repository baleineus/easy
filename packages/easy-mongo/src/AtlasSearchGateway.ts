import { stages } from './Stages';
import { MongoGateway } from './MongoGateway';
import { Collection } from './Collection';
import { MongoProvider } from './MongoProvider';
import { lucene, SearchDefinition } from './Lucene';
import { entries, Filter, Json, PageList, text, toPageList, tuple2 } from '@thisisagile/easy';

const { skip, take } = stages.skip;
const { replaceWith } = stages.replaceWith;
const { searchWithDef, searchMeta, facets } = lucene;

type FacetResult = { _id: string; count: number };
type Value = { label: string; value: string; count: number };
export const toFilters = (facets: unknown): Filter[] =>
  entries(facets as Record<string, unknown>).map(([k, fs]) => ({
    label: text(k).title.toString(),
    field: k,
    values: (fs as []).map((f: FacetResult) => toValue(f)),
  }));

const toValue = (f: FacetResult): Value => ({ label: f._id ?? 'unknown', value: f._id ?? 'unknown', count: f.count });

export class AtlasSearchGateway extends MongoGateway {
  constructor(
    collection: Collection,
    provider: MongoProvider = collection.provider,
    readonly searchDef: SearchDefinition,
    readonly sortDef: Record<string, Record<string, 1 | -1>> = {}
  ) {
    super(collection, provider);
  }

  query = async (query: Record<keyof typeof this.searchDef, string | number>): Promise<PageList<Json>> => {
    return tuple2(
      this.aggregate(searchWithDef(query, this.searchDef), skip({ skip: (query?.skip as number) ?? 0 }), take({ take: (query?.take as number) ?? 250 })),
      this.aggregate(
        searchMeta(query, this.searchDef),
        replaceWith({
          total: '$count.lowerBound',
          facets: Object.keys(facets(this.searchDef)).reduce((acc, k) => ({ ...acc, [k]: `$facet.${k}.buckets` }), {}),
        })
      )
    )
      .then(([data, meta]) => ({ data, meta: meta.first() as any }))
      .then(({ data, meta }) =>
        toPageList<Json>(data, {
          total: meta?.total ?? 0,
          skip: query?.skip as number,
          take: query?.take as number,
          sorts: Object.keys(this.sortDef),
          filters: toFilters(meta.facets),
        })
      );
  };
}
