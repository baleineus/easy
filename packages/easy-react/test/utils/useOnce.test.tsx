import React from 'react';
import '@thisisagile/easy-test';
import { rendersWait } from '@thisisagile/easy-test-react';
import { resolve } from '@thisisagile/easy';
import { useOnce } from '../../src';

describe('useOnce', () => {
  const Club = ({ f, initial }: { f: () => Promise<string>; initial?: string }) => {
    const [name] = useOnce(f, { initial });
    return <div data-testid={name}>Club</div>;
  };

  test('useOnce works.', async () => {
    const { container, atId } = await rendersWait(<Club f={() => resolve('barcelona')} />);
    expect(container).toMatchSnapshot();
    expect(atId('barcelona')).toBeValid();
  });

  test('useOnce works with initial.', async () => {
    const { container, atId } = await rendersWait(<Club f={() => resolve('barcelona')} initial="feyenoord" />);
    expect(container).toMatchSnapshot();
    expect(atId('barcelona')).toBeValid();
  });

  test('useOnce works with initial and undefined.', async () => {
    const { container, atId } = await rendersWait(<Club f={() => resolve(undefined as unknown as string)} initial={'feyenoord'} />);
    expect(container).toMatchSnapshot();
    expect(atId('feyenoord')).toBeValid();
  });
});
