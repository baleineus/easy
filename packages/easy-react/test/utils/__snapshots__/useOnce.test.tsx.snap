// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useOnce useOnce works with initial and reject. 1`] = `
<div>
  <div
    data-testid="feyenoord"
  >
    Club
  </div>
</div>
`;

exports[`useOnce useOnce works with initial and undefined. 1`] = `
<div>
  <div
    data-testid="feyenoord"
  >
    Club
  </div>
</div>
`;

exports[`useOnce useOnce works with initial. 1`] = `
<div>
  <div
    data-testid="barcelona"
  >
    Club
  </div>
</div>
`;

exports[`useOnce useOnce works. 1`] = `
<div>
  <div
    data-testid="barcelona"
  >
    Club
  </div>
</div>
`;
